### PR TITLE
[Keyboard] Update ZSA Moonlander

### PR DESCRIPTION
**Pull request by drashna**
## Description

* Updates matrix code to use mcp2301x, in line with voyager code
* Update kb code to be in line with voyager code (removing threading code!)
* Move a lot of config into keyboard.json
* At some point, chips were switched to STM32F303xB instead of xC.  Update ld script to reflect this.

## Types of Changes

- [x] Enhancement/optimization
- [x] Keyboard (addition or update)

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/zsa/moonlander/matrix.c [hunk 1]
```diff
@@ -14,62 +14,39 @@
...
+#define MCP23_ROW_PINS { GPB5, GBP4, GBP3, GBP2, GBP1, GBP0 }       outputs
+#define MCP23_COL_PINS { GPA0, GBA1, GBA2, GBA3, GBA4, GBA5, GBA6 } inputs
+
  */
 /* matrix state(1:on, 0:off) */
-extern matrix_row_t matrix[MATRIX_ROWS];      // debounced values
-extern matrix_row_t raw_matrix[MATRIX_ROWS];  // raw values
+extern matrix_row_t matrix[MATRIX_ROWS];     // debounced values
+extern matrix_row_t raw_matrix[MATRIX_ROWS]; // raw values
 static matrix_row_t raw_matrix_right[MATRIX_COLS];
```
Comment (zvecr) [inline keyboards/zsa/moonlander/matrix.c:34]: Use of `MATRIX_COLS` here looks odd?

File: keyboards/zsa/moonlander/matrix.c [hunk 2]
```diff
@@ -87,115 +64,117 @@ void matrix_init_custom(void) {
...
-            // data = 0x01;
-        } else {
-            data = 0;
-        }
-
-        if (raw_matrix_right[row] != data) {
-            raw_matrix_right[row] = data;
-            changed         = true;
+            mcp23018_errors += !mcp23018_set_output(MCP23018_DEFAULT_ADDRESS, mcp23018_PORTA, (0b01111111 & ~(1 << (row))) | ((uint8_t)!mcp23018_leds[2] << 7));
+            mcp23018_errors += !mcp23018_set_output(MCP23018_DEFAULT_ADDRESS, mcp23018_PORTB, ((uint8_t)!mcp23018_leds[1] << 6) | ((uint8_t)!mcp23018_leds[0] << 7));
```
Comment (zvecr) [inline keyboards/zsa/moonlander/matrix.c:120]: Probably want to use `mcp23018_set_output_all` here for a small speed boost?

File: keyboards/zsa/moonlander/matrix.c [hunk 3]
```diff
@@ -87,115 +64,117 @@ void matrix_init_custom(void) {
...
-                changed                = true;
+                current_matrix[row] = data;
+                changed             = true;
             }
         }
+
+        // Reading the right side of the keyboard.
+        if (!mcp23018_errors) {
+            uint8_t rx;
+            mcp23018_errors += !mcp23018_readPins(MCP23018_DEFAULT_ADDRESS, mcp23018_PORTB, &rx);
```
Comment (zvecr) [inline keyboards/zsa/moonlander/matrix.c:165]:
```suggestion
            mcp23018_errors += !mcp23018_read_pins(MCP23018_DEFAULT_ADDRESS, mcp23018_PORTB, &rx);
```

File: keyboards/zsa/moonlander/matrix.c [hunk 4]
```diff
@@ -14,62 +14,39 @@
...
-
-        if (MSG_OK != i2c_transmit(MCP23018_DEFAULT_ADDRESS << 1, mcp23018_tx, 3, MOONLANDER_I2C_TIMEOUT)) {
-            dprintf("error hori\n");
-        } else {
-            mcp23018_initd = is_launching = true;
-        }
-    }
+bool io_expander_ready(void) {
+    uint8_t tx;
+    return mcp23018_readPins(MCP23018_DEFAULT_ADDRESS, mcp23018_PORTA, &tx);
```
Comment (zvecr) [inline keyboards/zsa/moonlander/matrix.c:46]:
```suggestion
    return mcp23018_read_pins(MCP23018_DEFAULT_ADDRESS, mcp23018_PORTA, &tx);
```

File: keyboards/zsa/moonlander/matrix.c (hunk 2 above)
Comment (drashna) [reply keyboards/zsa/moonlander/matrix.c:120]: any speed boost is good.

**Review by zvecr — APPROVED** [review]

**Pull request merged**